<a href="https://colab.research.google.com/github/aaronjoseph/KB_Final/blob/master/Decoding_Roberta.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [4]:
!pip install transformers

In [6]:
from transformers import RobertaConfig, RobertaModel

In [9]:
# Initializing a RoBERTa configuration
configuration = RobertaConfig()
# Initializing a model from the configuration
model = RobertaModel(configuration)
# Accessing the model configuration
configuration = model.config

In [12]:
from torch import cuda
device = 'cuda' if cuda.is_available() else 'cpu'
model.to(device)

RobertaModel(
  (embeddings): RobertaEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=1)
    (position_embeddings): Embedding(512, 768, padding_idx=1)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): RobertaEncoder(
    (layer): ModuleList(
      (0): RobertaLayer(
        (attention): RobertaAttention(
          (self): RobertaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): RobertaSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Drop

In [13]:
configuration

RobertaConfig {
  "attention_probs_dropout_prob": 0.1,
  "bos_token_id": 0,
  "eos_token_id": 2,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "roberta",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 1,
  "position_embedding_type": "absolute",
  "transformers_version": "4.3.3",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
}

## attention_probs_dropout_prob : 0.1
Is defined as the attention probabilites

## bos_token_id : 0
The id of the beginning-of-sequence token

## eos_token_id : 2
The id of the end-of-sequence token

## gradient_checkpointing: false

This can be used to save GPU memory

Gradient checkpointing lets you fit 10x larger neural nets into memory at the cost of an additional 20% computation time. The tools within this package, which is a joint development of Tim Salimans and Yaroslav Bulatov, aids in rewriting TensorFlow model for using less memory.

[Links](https://github.com/huggingface/transformers/issues/6258)

[Links](https://hub.packtpub.com/openais-gradient-checkpointing-package-makes-huge-neural-nets-fit-memory/#:~:text=Gradient%20checkpointing%20lets%20you%20fit,model%20for%20using%20less%20memory.)

## "hidden_act" : gelu

### GELU : Gaussian Error Linear Unit

GELU(x)=xP(X≤x)=xΦ(x)  
$$= 0.5x(1+tanh[(2/π)^{0.5} (x+0.044715x3)])$$

[GELU Link](https://datascience.stackexchange.com/questions/49522/what-is-gelu-activation)

## "hidden_dropout_prob": 0.1,

The dropout probability for all fully connected layers in the embeddings, encoder, and pooler

## "hidden_size": 768

The number of hidden layers

## "initializer_range": 0.02,

Initializers define the way to set the initial random weights of model layers

## intermediate_size

Dimensionality of the “intermediate” (often named feed-forward) layer in the Transformer encoder

## layer_norm_eps 

The epsilon used by the layer normalization layers

## "max_position_embeddings": 512

The maximum sequence length that this model might ever be used with. Typically set this to something large just in case (e.g., 512 or 1024 or 2048).

## "model_type": "roberta"

Type of model

## "num_attention_heads": 12

Number of attention heads for each attention layer in the Transformer encoder

## "num_hidden_layers": 12

Number of hidden layers in the Transformer encoder.

## "pad_token_id": 1

A special token used to make arrays of tokens the same size for batching purpose. Will then be ignored by attention mechanisms or loss computation.

[Values](https://github.com/huggingface/transformers/issues/6238)

## "position_embedding_type": "absolute"

Type of position embedding. Choose one of "absolute", "relative_key", "relative_key_query". For positional embeddings use "absolute".

[Research Paper](https://arxiv.org/abs/2009.13658)

## "transformers_version": "4.3.3",

Version Number of the module

## "type_vocab_size": 2

 The vocabulary size of the token_type_ids passed into Roberta

## "use_cache": true

If set to True, past_key_values key value states are returned and can be used to speed up decoding 

## "vocab_size": 30522

RoBERTa uses a Byte-Level BPE tokenizer with a larger subword vocabulary